lib/internal/test_runner/tap_parser.js

```diff
@@ -0,0 +1,827 @@
+  // ----------------------------------------------------------------------//
+
+  async parseAsync(chunk, callback = null) {
+    this.#isAsyncParsingEnabled = true;
+    this.#currentTokenChunk = 0;
+    this.#currentTokenIndex = 0;
+    this.input = chunk;
+    this.#lexer = new TapLexer(chunk);
+
+    try {
+      this.#tokens = this.#makeChunks(this.#lexer.scan());
+      this.#parseTokens(callback);
+    } catch (error) {
+      callback(null, error);
+    }
+  }
+
+  parse(input, callback = null) {
+    this.#isAsyncParsingEnabled = false;
+    this.input = input;
+    this.#lexer = new TapLexer(input);
+    this.#tokens = this.#makeChunks(this.#lexer.scan());
+    this.#parseTokens(callback);
+
+    if (this.#isYAMLBlock) {
+      // Looks like we have a non-ending YAML block
+      this.#error('Expected end of YAML block', this.#tokens.at(-1).at(-1));
+    }
+
+    this.#output = {
+      root: { ...this.#documents },
+    };
+
+    return this.#output;
+  }
+
+  // Check if the TAP content is semantically valid
+  check() {
+    return this.#checker.check(this.#output);
+  }
+
+  // ----------------------------------------------------------------------//
+  // --------------------------- Transform API ----------------------------//
+  // ----------------------------------------------------------------------//
+
+  _transform(chunk, _encoding, next) {
+    const line = chunk.toString('utf-8');
+    this.parseAsync(line, (ast, error) => {
+      if (error) {
+        this.emit('error', error);
+      } else {
+        this.push({
+          __proto__: null,
+          ...ast,
+        });
+      }
+    });
+    next();
+  }
+
+  // ----------------------------------------------------------------------//
+  // ----------------------------- Private API ----------------------------//
+  // ----------------------------------------------------------------------//
+
+  #parseTokens(callback = null) {
+    this.#tokens.forEach((chunk) => this.#parseChunk(chunk, callback));
+  }
+
+  #parseChunk(chunk, callback = null) {
+    this.#subTestLevel = this.#getCurrentIndentationLevel();
+    // If the chunk starts with a multiple of 4 whitespace tokens,
+    // then it's a subtest
+    if (this.#subTestLevel > 0) {
+      // Remove the indentation block from the current chunk
+      // but only if the chunk is not a YAML block
+
+      if (!this.#isYAMLBlock) {
+        chunk = chunk.slice(
+          this.#subTestLevel * this.#kSubtestBlockIndentationFactor
+        );
+      }
+    }
+
+    const astNode = this.#TAPDocument(chunk);
+
+    callback?.({
+      ...astNode,
+      level: this.#subTestLevel,
+      lexeme: chunk.map((token) => token.value).join(''),
+      tokens: chunk,
+    });
+
+    // Move pointers to the next chunk and reset the current token index
+    this.#currentTokenChunk++;
+    this.#currentTokenIndex = 0;
+  }
+
+  #error(message, token, received = null) {
```

Comment:
we should not throw when failing to parse - instead, the entire line should become a `TokenKind.COMMENT`

Reply:
Parsing won't fail when running in a stream context.